pkg/oci/routes.go

```diff
@@ -0,0 +1,291 @@
+
+Ultimately the reconciliation restores order in the galaxy, but only after some
+disruptions in the force.
+
+One _could_ add some locking on the Get/Update of RouteTable, but it's not
+clear that should be required.
+*/
+
+// CreateRoute for OCI, we have to udpate the whole RouteTable not just add the singular route
+func (cp *CloudProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	glog.V(6).Info("Adding route to route table: ", route)
+
+	node, err := cp.NodeLister.Get(string(route.TargetNode))
+
+	ocid := util.MapProviderIDToInstanceID(node.Spec.ProviderID)
+	vnic, err := cp.client.Compute().GetPrimaryVNICForInstance(ctx, "", ocid)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	ipOcid, err := cp.client.Networking().GetOCIDFromIP(ctx, *vnic.PrivateIp, *vnic.SubnetId)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	glog.V(6).Infof("ipOcid: ", ipOcid)
+
+	for routeTableID := range routeTableIds {
+		routeTable, err := cp.client.Networking().GetRouteTable(ctx, routeTableID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		for _, oRoute := range routeTable.RouteRules {
+			if *oRoute.CidrBlock == route.DestinationCIDR {
+				if *oRoute.NetworkEntityId == ipOcid {
+					// skipping existing route table entry
+					glog.V(6).Info("Skipping existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					continue
+				} else {
+					oRoute.NetworkEntityId = &ipOcid
+					glog.V(6).Info("Update existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					err := cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+					if err != nil {
+						return errors.WithStack(err)
+					}
+					continue
+				}
+			}
+		}
+
+		// route rule doesn't exist in RouteTable
+		glog.V(6).Info("Add new route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+		routeTable.RouteRules = append(routeTable.RouteRules, core.RouteRule{
+			CidrBlock:       &route.DestinationCIDR,
+			NetworkEntityId: &ipOcid,
+		})
+
+		err = cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+// DeleteRoute delete a route entry, but OCI only has Update of the entire RouteTable
+func (cp *CloudProvider) DeleteRoute(ctx context.Context, clusterName string, route *cloudprovider.Route) error {
+	glog.V(6).Infof("Delete Route: %v", route)
+
+	for routeID := range routeTableIds {
+		oRoute, err := cp.client.Networking().GetRouteTable(ctx, routeID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		dirty := false
+
+		for i, r := range oRoute.RouteRules {
+			if *r.CidrBlock != route.DestinationCIDR {
+				continue
+			}
+
+			dirty = true
+
+			glog.V(6).Infof("Deleting Route: %s %s", route.DestinationCIDR, route.TargetNode)
+			oRoute.RouteRules = append(oRoute.RouteRules[:i], oRoute.RouteRules[i+1:]...)
```

Comment:
This will panic when you remove the last rule in the list since i+1 isn't an index.